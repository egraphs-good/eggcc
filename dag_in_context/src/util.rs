use std::{
    ffi::OsStr,
    fs::{self, File},
    process::{Command, Stdio},
    time::{SystemTime, UNIX_EPOCH},
};

/// Invokes some program with the given arguments, piping the given input to the program.
/// Returns an error if the program returns a non-zero exit code.
/// Code adapted from https://github.com/egraphs-good/egg/blob/e7845c5ae34267256b544c8e6b5bc36d91d096d2/src/dot.rs#L127
pub fn run_cmd_line<S1, S2, I>(program: S1, args: I, input: &str) -> std::io::Result<String>
where
    S1: AsRef<OsStr>,
    S2: AsRef<OsStr>,
    I: IntoIterator<Item = S2>,
{
    // Write the input to a temporary file so the child can read it directly.
    let mut temp_path = std::env::temp_dir();
    let unique = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_nanos();
    temp_path.push(format!(
        "eggcc-run-cmd-{}-{}.tmp",
        std::process::id(),
        unique
    ));

    fs::write(&temp_path, input)?;
    let input_file = File::open(&temp_path)?;

    let output = Command::new(program)
        .args(args)
        .stdin(Stdio::from(input_file))
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output();

    let output = match output {
        Ok(out) => {
            let _ = fs::remove_file(&temp_path);
            out
        }
        Err(err) => {
            let _ = fs::remove_file(&temp_path);
            return Err(err);
        }
    };

    match output.status.code() {
        Some(0) => Ok(String::from_utf8(output.stdout)
            .map_err(|e| std::io::Error::other(format!("utf8 error: {}", e)))?),
        Some(exit) => {
            let stderr = String::from_utf8_lossy(&output.stderr);
            Err(std::io::Error::other(format!(
                "program returned error code {exit}; stderr: {stderr}"
            )))
        }
        None => {
            let stderr = String::from_utf8_lossy(&output.stderr);
            Err(std::io::Error::other(format!(
                "program was terminated by signal; stderr: {stderr}"
            )))
        }
    }
}
