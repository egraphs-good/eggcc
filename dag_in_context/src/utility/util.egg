(function ListExpr-length (ListExpr) i64)
(function ListExpr-ith (ListExpr i64) Expr :unextractable)
(function ListExpr-suffix (ListExpr i64) ListExpr :unextractable)
(function Append (ListExpr Expr) ListExpr :unextractable)

(rule ((Switch pred inputs branch)) ((union (ListExpr-suffix branch 0) branch)) :ruleset always-run)

(rule ((= (ListExpr-suffix top n) (Cons hd tl)))
    ((union (ListExpr-ith top n) hd)
     (union (ListExpr-suffix top (+ n 1)) tl)) :ruleset always-run)

(rule ((= (ListExpr-suffix list n) (Nil)))
    ((set (ListExpr-length list) n)) :ruleset always-run)

(rewrite (Append (Cons a b) e)
   (Cons a (Append b e))
   :ruleset always-run)
(rewrite (Append (Nil) e)
   (Cons e (Nil))
   :ruleset always-run)

(function tuple-length (Expr) i64 :unextractable)

(rule ((HasType expr (TupleT tl))
       (= len (TypeList-length tl)))
      ((set (tuple-length expr) len)) :ruleset always-run)

;; Create a Get for every index, and rewrite it to see through Concat
(rule ((Single expr)) ((union (Get (Single expr) 0) expr)) :ruleset always-run)
;; initial get
(rule ((Concat expr1 expr2)
       (> (tuple-length (Concat expr1 expr2)) 0))
      ((Get (Concat expr1 expr2) 0))
      :ruleset always-run)
;; next get
(rule ((= concat (Concat expr1 expr2))
       (= len (tuple-length concat))
       (= ith (Get concat i))
       (< (+ i 1) len)
       )
       ((Get (Concat expr1 expr2) (+ 1 i)))
       :ruleset always-run)

;; descend left
(rule ((Get (Concat expr1 expr2) i)
       (= (tuple-length expr1) len1)
       (< i len1))
      ((union (Get (Concat expr1 expr2) i)
              (Get expr1 i)))
       :ruleset always-run)
;; descend right
(rule ((Get (Concat expr1 expr2) i)
       (= (tuple-length expr1) len1)
       (>= i len1))
      ((union (Get (Concat expr1 expr2) i)
              (Get expr2 (- i len1))))
       :ruleset always-run)


(relation IsLeaf (Expr))
(rule ((Arg ty)) ((IsLeaf (Arg ty))) :ruleset always-run)
(rule ((Const val ty)) ((IsLeaf (Const val ty))) :ruleset always-run)
(rule ((Empty ty)) ((IsLeaf (Empty ty))) :ruleset always-run)

(rewrite (Concat (Concat a b) c)
         (Concat a (Concat b c))
         :ruleset always-run)

;;                tuple start len
(function SubTuple (Expr i64 i64) Expr :unextractable)
(rewrite (SubTuple expr x 1)
         (Single (Get expr x)) :ruleset always-run)
(rule ((= f (SubTuple expr a b)) (> b 1))
      ((union f 
         (Concat (Single (Get expr a))
                 (SubTuple expr (+ a 1) (- b 1))))) :ruleset always-run)