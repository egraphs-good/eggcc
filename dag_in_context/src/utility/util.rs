#[cfg(test)]
use crate::{egglog_test, interpreter::Value};

#[test]
fn test_list_util() -> crate::Result {
    use crate::ast::emptyt;
    use crate::schema::Assumption;
    let emptyt = emptyt();
    let ctx = Assumption::dummy();
    let build = format!(
        "
		(let list (Cons (Const (Int 0) {emptyt} {ctx})
                  (Cons (Const (Int 1) {emptyt} {ctx})
                  (Cons (Const (Int 2) {emptyt} {ctx})
                  (Cons (Const (Int 3) {emptyt} {ctx})
                  (Cons (Const (Int 4) {emptyt} {ctx}) (Nil)))))))
		(let expr (Switch (Const (Int 1) {emptyt} {ctx}) (Empty {emptyt} {ctx}) list))
	"
    );
    let check = format!(
        "
		(check (= (ListExpr-ith list 1) (Const (Int 1) {emptyt} {ctx})))
        (check (= (ListExpr-ith list 4) (Const (Int 4) {emptyt} {ctx})))
        (check (= (ListExpr-length list) 5))
	"
    );
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}

#[test]
fn append_test() -> crate::Result {
    use crate::ast::emptyt;
    use crate::schema::Assumption;
    let emptyt = emptyt();
    let ctx = Assumption::dummy();
    let build = format!(
        "
        (let appended
            (Append
                (Cons (Const (Int 0) {emptyt} {ctx}) (Cons (Const (Int 1) {emptyt} {ctx}) (Nil)))
                (Const (Int 2) {emptyt} {ctx})))
    "
    );

    let check = format!("
        (check (
            =
            (Cons (Const (Int 0) {emptyt} {ctx}) (Cons (Const (Int 1) {emptyt} {ctx}) (Cons (Const (Int 2) {emptyt} {ctx}) (Nil))))
            appended
        ))
    ");
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}

#[test]
fn test_tuple_ith() -> crate::Result {
    use crate::ast::emptyt;
    use crate::schema::Assumption;
    let emptyt = emptyt();
    let ctx = Assumption::dummy();
    let build = format!(
        "
    (let tup (Concat
                  (Concat (Single (Const (Int 0) {emptyt} {ctx})) (Single (Const (Int 1) {emptyt} {ctx})))
                  (Concat (Single (Const (Int 2) {emptyt} {ctx})) (Single (Const (Int 3) {emptyt} {ctx})))))
    
    ;; Gets of tup and tup2 are only generated when tup and tup2 are in critical positions
    ;; like inputs to a DoWhile, 
    (let tup_type (TupleT (TCons (IntT) (TCons (IntT) (TCons (IntT) (TCons (IntT) (TNil)))))))
    (let body (Concat (Const (Bool true) tup_type (TmpCtx)) (Arg tup_type (TmpCtx))))
    (let loop (DoWhile tup body))
    (union (TmpCtx) (InLoop tup body))
    (delete (TmpCtx))
    
    ;; with print
    (let tup2 (Concat 
                (Concat 
                    (Single (Bop (Print) (Const (Int 0) {emptyt} {ctx}) (Arg (Base (StateT)) {ctx})))
                    (Concat (Single (Const (Int 1) {emptyt} {ctx})) 
                                (Single (Const (Int 2) {emptyt} {ctx}))))
                (Single (Const (Int 3) {emptyt} {ctx}))))
    (let tup2_type (TupleT (TCons (StateT) (TCons (IntT) (TCons (IntT) (TCons (IntT) (TNil)))))))
    (let body2 (Concat (Const (Bool true) tup2_type (TmpCtx)) (Arg tup2_type (TmpCtx))))
    (let loop2 (DoWhile tup2 body2))
    (union (TmpCtx) (InLoop tup2 body2))
    (delete (TmpCtx))
    "
    );

    let check = format!(
        "
    (check (= (Get tup 0) (Const (Int 0) {emptyt} {ctx})))
    (check (= (Get tup 1) (Const (Int 1) {emptyt} {ctx})))
    (check (= (Get tup 2) (Const (Int 2) {emptyt} {ctx})))
    (check (= (Get tup 3) (Const (Int 3) {emptyt} {ctx})))
    (check (= 4 (tuple-length tup)))
    (fail (check (Get tup 4)))

    (check (= (Get tup2 0) (Bop (Print) (Const (Int 0) {emptyt} {ctx}) (Arg (Base (StateT)) {ctx}))))
    (check (= (Get tup2 1) (Const (Int 1) {emptyt} {ctx})))
    (check (= (Get tup2 2) (Const (Int 2) {emptyt} {ctx})))
    (check (= (Get tup2 3) (Const (Int 3) {emptyt} {ctx})))
    (check (= 4 (tuple-length tup2)))
    (fail (check (Get tup2 4)))
    "
    );
    egglog_test(
        &build,
        &check,
        vec![],
        Value::Tuple(vec![]),
        Value::Tuple(vec![]),
        vec![],
    )
}
