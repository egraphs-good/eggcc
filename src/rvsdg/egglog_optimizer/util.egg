;; ####################################
;; implementation of PassThroughArguments
;; Creates a vec of arguments
;; (vec-of (Arg 0) (Arg 1) ...) with length i
(function PassThroughArguments (i64) VecOperand :unextractable)



;; (how many arguments to generate, vector so far)
(function PassThroughArgumentsHelper (i64 VecOperandBase) VecOperand :unextractable)

(rewrite (PassThroughArguments i)
  (PassThroughArgumentsHelper i (vec-of))
  :ruleset subst)

(rule ((= lhs (PassThroughArgumentsHelper i rest))
       (< (vec-length rest) i))
      ((union lhs
        (PassThroughArgumentsHelper i
            (vec-push rest (Arg (vec-length rest))))))
      :ruleset subst)

(rule ((= lhs (PassThroughArgumentsHelper i rest))
       (= (vec-length rest) i))
      ((union lhs (VO rest)))
      :ruleset subst)

;; Project each argument out of a body
(function BodyToVecOperand (i64 Body) VecOperand :unextractable)
;; current index, body length, body, and vector so far
(function BodyToVecOperandHelper (i64 i64 Body VecOperandBase) VecOperand :unextractable)

(rewrite (BodyToVecOperand body-len body)
         (BodyToVecOperandHelper 0 body-len body (vec-of)) :ruleset subst)
(rule
  ((= helper (BodyToVecOperandHelper index body-len body so-far))
   (< index body-len))
  ((union helper
          (BodyToVecOperandHelper (+ index 1) body-len body
                                  (vec-push so-far 
                                   (Project index body)))))
  :ruleset subst)

(rule
  ((= helper (BodyToVecOperandHelper index body-len body so-far))
   (= index body-len))
  ((union helper (VO so-far)))
  :ruleset subst)


