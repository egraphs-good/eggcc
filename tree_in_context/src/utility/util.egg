(function ListExpr-length (ListExpr) i64)
(function ListExpr-ith (ListExpr i64) Expr :unextractable)
(function ListExpr-suffix (ListExpr i64) ListExpr :unextractable)
(relation DemandListExpr-ith (ListExpr))
(function Append (ListExpr Expr) ListExpr :unextractable)

(rule ((Switch pred branch)) ((union (ListExpr-suffix branch 0) branch)) :ruleset always-run)


(rule ((DemandListExpr-ith list))
      ((union (ListExpr-suffix list 0) list)) :ruleset always-run)
(rule ((= (ListExpr-suffix top n) (Cons hd tl)))
    ((union (ListExpr-ith top n) hd)
     (union (ListExpr-suffix top (+ n 1)) tl)) :ruleset always-run)

(rule ((= (ListExpr-suffix list n) (Nil)))
    ((set (ListExpr-length list) n)) :ruleset always-run)

(rewrite (Append (Cons a b) e)
   (Cons a (Append b e))
   :ruleset always-run)
(rewrite (Append (Nil) e)
   (Cons e (Nil))
   :ruleset always-run)

(function tuple-ith (Expr i64) Expr :unextractable)
(function tuple-length (Expr) i64 :unextractable)

(rule ((HasType expr (TupleT tl))
       (= len (TypeList-length tl)))
      ((set (tuple-length expr) len)) :ruleset always-run)

(rule ((Single expr)) ((union (tuple-ith (Single expr) 0) expr)) :ruleset always-run)
(rule ((Concat ord expr1 expr2)
       (= ith (tuple-ith expr1 i)))
       ((union (tuple-ith (Concat ord expr1 expr2) i) ith)) :ruleset always-run)

(rule ((Concat ord expr1 expr2)
       (= len1 (tuple-length expr1))
       (= ith (tuple-ith expr2 i)))
       ((union (tuple-ith (Concat ord expr1 expr2) (+ len1 i)) ith)) :ruleset always-run)


(sort ListAssume)
(function AssumeNil () ListAssume)
(function AssumeCons (Assume ListAssume) ListAssume)

(relation DemandListAssume-ith (ListAssume))

(function ListAssume-suffix (ListAssume i64) ListAssume :unextractable)
(function ListAssume-ith (ListAssume i64) Assume :unextractable)

(rule ((DemandListAssume-ith list))
      ((union (ListAssume-suffix list 0) list)) :ruleset always-run)

(rule ((= (ListAssume-suffix top n) (AssumeCons hd tl)))
    ((union (ListAssume-ith top n) hd)
     (union (ListAssume-suffix top (+ n 1)) tl)) :ruleset always-run)
