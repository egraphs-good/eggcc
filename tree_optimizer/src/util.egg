(function ListExpr-length (ListExpr) i64)
(function ListExpr-ith (ListExpr i64) Expr :unextractable)
(function ListExpr-suffix (ListExpr i64) ListExpr :unextractable)
(function Append (ListExpr Expr) ListExpr :unextractable)

(rule ((All id order top)) ((union (ListExpr-suffix top 0) top)) :ruleset always-run)

(rule ((= (ListExpr-suffix top n) (Cons hd tl)))
    ((union (ListExpr-ith top n) hd)
     (union (ListExpr-suffix top (+ n 1)) tl)) :ruleset always-run)

(rule ((= (ListExpr-suffix list n) (Nil)))
    ((set (ListExpr-length list) n)) :ruleset always-run)

(rewrite (Append (Cons a b) e)
    (Cons a (Append b e))
    :ruleset always-run)
(rewrite (Append (Nil) e)
    (Cons e (Nil))
    :ruleset always-run)

(function Expr-size (Expr) i64 :merge (min old new))
(function ListExpr-size (ListExpr) i64 :merge (min old new))
