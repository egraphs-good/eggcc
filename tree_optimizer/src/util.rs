use crate::{
    expr::Expr,
    ir::{Constructor, Purpose},
};
use std::iter;

fn ast_size_for_ctor(ctor: Constructor) -> String {
    let ctor_pattern = ctor.construct(|field| field.var());
    let ruleset = " :ruleset always-run";
    match ctor {
        // List itself don't count size
        Constructor::Nil =>  format!("(rule ({ctor_pattern}) ((set (ListExpr-size {ctor_pattern}) 0)) {ruleset})"),
        Constructor::Cons => format!("(rule ((= list (Cons expr xs)) (= a (Expr-size expr)) (= b (ListExpr-size xs))) ((set (ListExpr-size list) (+ a b))){ruleset})"), 
        // let Get and All's size = children's size (I prefer not +1 here)
        Constructor::Expr(Expr::Get(..)) => format!("(rule ((= expr (Get tup i)) (= n (Expr-size tup))) ((set (Expr-size expr) n)) {ruleset})"),
        Constructor::Expr(Expr::All(..)) => format!("(rule ((= expr (All id ord list)) (= n (ListExpr-size list))) ((set (Expr-size expr) n)) {ruleset})"),
        Constructor::Expr(Expr::Branch(..)) => format!("
(rule ((= expr (Branch id child))
       (= n (Expr-size child)))
      ((set (Expr-size expr) n)) {ruleset})
        "),
        _ => {
            let field_pattern = ctor.filter_map_fields(|field| {
                let sort = field.sort().name();
                let var = field.var();
                match field.purpose {
                    Purpose::CapturedExpr
                    | Purpose::SubExpr
                    | Purpose::SubListExpr =>
                        Some(format!("({sort}-size {var})")),
                    _ => None
                }
            });

            let result_str = sum_vars(field_pattern);
            format!("(rule ((= expr {ctor_pattern}) (= sum {result_str})) ((set (Expr-size expr) (+ 1 sum))){ruleset})")
        },
    }
}

fn sum_vars(vars: Vec<String>) -> String {
    let len = vars.len();
    match len {
        0 => "0".to_string(),
        1 => vars[0].clone(),
        2 => format!("(+ {} {})", vars[0], vars[1]),
        3 => format!("(+ {} (+ {} {}))", vars[0], vars[1], vars[2]),
        _ => panic!("Unimplemented"),
    }
}

pub(crate) fn rules() -> Vec<String> {
    iter::once(include_str!("util.egg").to_string())
        .chain(Constructor::iter().map(ast_size_for_ctor))
        .collect::<Vec<_>>()
}

#[test]
fn test_list_util() -> crate::Result {
    let build = &*"
        (let id (Id 1))
        (let list (Cons (Num id 0) (Cons (Num id 1) (Cons (Num id 2) (Cons (Num id 3) (Cons (Num id 4) (Nil)))))))
        (let t (All id (Sequential) list))
    ".to_string();
    let check = &*"
        (check (= (ListExpr-ith list 1) (Num id 1)))
        (check (= (ListExpr-ith list 4) (Num id 4)))
        (check (= (ListExpr-length list) 5))
        
    "
    .to_string();
    crate::run_test(build, check)
}

#[test]
fn append_test() -> crate::Result {
    let build = "
        (let id (Id (i64-fresh!)))
        (let appended
            (Append
                (Cons (Num id 0) (Cons (Num id 1) (Nil)))
                (Num id 2)))
    ";

    let check = "
        (check (
            =
            (Cons (Num id 0) (Cons (Num id 1) (Cons (Num id 2) (Nil))))
            appended
        ))
    ";

    crate::run_test(build, check)
}

#[test]
fn ast_size_test() -> crate::Result {
    let build = "
    (let inv 
        (BOp (Sub) (Get (Arg shared) 4)
            (BOp (Mul) (Get (Arg shared) 2) 
                (Switch (Num shared 1)
                  (list4 (Branch shared (Num shared 1))
                         (Branch shared (Num shared 2))
                         (Branch shared (Num shared 3))
                         (Branch shared (Num shared 4)))
                )
            )
        ))
    
    (let loop
        (Loop shared
            (All shared (Parallel) (list5 (Num shared 0)
                                    (Num shared 1)
                                    (Num shared 2)
                                    (Num shared 3)
                                    (Num shared 4)))
            (All shared (Sequential)
                (Pair
                ; pred
                (BOp (LessThan) (Get (Arg shared) 0) (Get (Arg shared) 4))
                ; output
                (All shared (Parallel) 
                    (list5
                        (BOp (Add) (Get (Arg shared) 0) 
                            inv
                        )
                        (Get (Arg shared) 1)
                        (Get (Arg shared) 2)
                        (Get (Arg shared) 3)
                        (Get (Arg shared) 4) ))))))
    ";

    let check = "
       (check (= 10 (Expr-size inv)))
       (check (= 25 (Expr-size loop)))
    ";

    crate::run_test(build, check)
}
