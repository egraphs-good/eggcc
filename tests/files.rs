use std::collections::HashSet;

use eggcc::util::{InterpMode, Run, RunType, TestProgram};
use insta::assert_snapshot;
use libtest_mimic::Trial;

/// Generate tests for all configurations of a given file
/// If `just_brilift` is true, only generate tests that
/// run the full pipeline with brilift
fn generate_tests(glob: &str, benchmark_mode: bool) -> Vec<Trial> {
    let mut trials = vec![];

    let mut mk_trial = |run: Run, snapshot: bool| {
        let snapshot_configurations: HashSet<RunType> = [RunType::Optimize].into_iter().collect();

        trials.push(Trial::test(run.name(), move || {
            let result = match run.run() {
                Err(error) => {
                    panic!("{}", error);
                }
                Ok(res) => res,
            };
            if run.test_type == RunType::CompileBrilift || run.test_type == RunType::CompileBrilLLVM
            {
                let executable = run
                    .output_path
                    .clone()
                    .unwrap_or_else(|| format!("/tmp/{}", run.name()));
                std::process::Command::new("rm")
                    .args(vec![executable.clone(), executable + "-args"])
                    .status()
                    .unwrap();
            }

            if run.interp.should_interp()
                && result.original_interpreted.as_ref().unwrap()
                    != result.result_interpreted.as_ref().unwrap()
            {
                panic!(
                    "Interpreted result does not match expected:\nExpected: {}\nGot: {}",
                    result.original_interpreted.unwrap(),
                    result.result_interpreted.unwrap()
                );
            }
            // only assert a snapshot if we are in the "small" folder
            if snapshot && snapshot_configurations.contains(&run.test_type) {
                for visualization in result.visualizations {
                    assert_snapshot!(run.name() + &visualization.name, visualization.result);
                }
            }
            Ok(())
        }))
    };

    for entry in glob::glob(glob).unwrap() {
        let f = entry.unwrap();

        let snapshot = f.to_str().unwrap().contains("small");

        // in benchmark mode, run fewer tests and use binaries to interpret
        let configurations = if benchmark_mode {
            // use InterpFast, which runs brilift
            // with optimization but without egglog optimization
            vec![
                Run::compile_brilift_config(
                    TestProgram::BrilFile(f.clone()),
                    true,
                    true,
                    InterpMode::InterpFast,
                ),
                Run::compile_llvm_config(
                    TestProgram::BrilFile(f.clone()),
                    true,
                    true,
                    InterpMode::InterpFast,
                ),
                Run::compile_llvm_config(
                    TestProgram::BrilFile(f),
                    false,
                    true,
                    InterpMode::InterpFast,
                ),
            ]
        } else {
            Run::all_configurations_for(TestProgram::BrilFile(f))
        };

        for run in configurations {
            mk_trial(run, snapshot);
        }
    }

    trials
}

fn main() {
    let args = libtest_mimic::Arguments::from_args();
    let mut tests = generate_tests("tests/passing/**/*.bril", false);
    // also generate tests for benchmarks
    tests.extend(generate_tests("benchmarks/passing/**/*.bril", true));

    libtest_mimic::run(&args, tests).exit();
}
