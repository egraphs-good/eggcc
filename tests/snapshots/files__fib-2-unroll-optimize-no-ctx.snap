---
source: tests/files.rs
expression: visualization.result
---
# ARGS: 100000
@main(v0: int) {
  v1_: ptr<int> = alloc v0;
  c2_: int = const 1;
  v3_: ptr<int> = ptradd v1_ c2_;
  c4_: int = const 0;
  store v1_ c4_;
  store v3_ c2_;
  c5_: int = const 2;
  v6_: ptr<int> = id v1_;
  v7_: int = id c5_;
  v8_: int = id c2_;
  v9_: int = id v0;
.b10_:
  v11_: bool = lt v7_ v9_;
  c12_: bool = const false;
  v13_: ptr<int> = id v6_;
  v14_: int = id v7_;
  v15_: bool = id c12_;
  v16_: int = id v8_;
  v17_: int = id v9_;
  br v11_ .b18_ .b19_;
.b18_:
  v20_: int = add v7_ v8_;
  v21_: bool = lt v20_ v9_;
  v22_: ptr<int> = ptradd v6_ v7_;
  c23_: int = const -2;
  v24_: ptr<int> = ptradd v22_ c23_;
  c25_: int = const -1;
  v26_: ptr<int> = ptradd v22_ c25_;
  v27_: int = load v26_;
  v28_: int = load v24_;
  v29_: int = add v27_ v28_;
  store v22_ v29_;
  v30_: ptr<int> = id v6_;
  v31_: int = id v20_;
  v32_: int = id v8_;
  v33_: int = id v9_;
  br v21_ .b34_ .b35_;
.b34_:
  v36_: ptr<int> = ptradd v6_ v20_;
  c37_: int = const -2;
  v38_: ptr<int> = ptradd v36_ c37_;
  c39_: int = const -1;
  v40_: ptr<int> = ptradd v36_ c39_;
  v41_: int = load v40_;
  v42_: int = load v38_;
  v43_: int = add v41_ v42_;
  store v36_ v43_;
  v44_: int = add v20_ v8_;
  v30_: ptr<int> = id v6_;
  v31_: int = id v44_;
  v32_: int = id v8_;
  v33_: int = id v9_;
.b35_:
  v45_: int = add v20_ v8_;
  v46_: int = select v21_ v45_ v20_;
  v13_: ptr<int> = id v6_;
  v14_: int = id v46_;
  v15_: bool = id v21_;
  v16_: int = id v8_;
  v17_: int = id v9_;
.b19_:
  v47_: int = add v7_ v8_;
  v48_: bool = lt v47_ v9_;
  c49_: bool = const false;
  v50_: bool = select v11_ v48_ c49_;
  v6_: ptr<int> = id v6_;
  v7_: int = id v14_;
  v8_: int = id v8_;
  v9_: int = id v9_;
  br v50_ .b10_ .b51_;
.b51_:
  v52_: ptr<int> = ptradd v6_ v7_;
  c53_: int = const -1;
  v54_: ptr<int> = ptradd v52_ c53_;
  v55_: int = load v54_;
  free v6_;
  print v55_;
  ret;
}
